File: sach/de3346.py
//DE3346.PY - Tran Huu Nam - hsgtin.vn - 2024
import bisect
n = int(input())
a = [0] * 1002
for i in range(1, n + 1):
    a[i] = int(input())
a[1:n+1] = sorted(a[1:n+1])
for i in range(1, n + 1):
    a[i] *= a[i]
dem = 0
for i in range(1, n - 1):
    for j in range(i + 1, n):
        k = a[i] + a[j]
        d = bisect.bisect_left(a[j+1:n+1], k) + j + 1
        while d <= n and a[d] == k:
            dem += 1
            d += 1
print(dem)

File: sach/de3346.cpp
//DE3346.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
long long n, a[1002];
int main() {
    cin>>n;
    int i,j, d; long long k, dem=0;
    for (i=1; i<=n; i++) cin>>a[i];
    sort(a+1, a+n+1);
    for (i=1; i<=n; i++) a[i]*=a[i];
    for (i=1; i<=n-2; i++)
        for (j=i+1; j<=n-1; j++) {
            k=a[i]+a[j];
            d=lower_bound(a+j+1, a+n+1, k) - a;
            while (d<=n && a[d]==k) {
                dem++;
                d++;
            }
        }
    cout<<dem;
}

